### ISSUE TITLE
Support exporting .sav files

### ISSUE THREAD

**Issue by jbmorley** [jbmorley/gameplay#46]

**Comment by jbmorley** [comment]
@taisel Do you know if any of the existing GameBoy-Online save methods support saving in a .sav format compatible with other emulators? There look to be quite a few different formats supported but I'm not too sure what I should be looking for... Thanks!

**Comment by taisel** [comment]
See the exporting as "RAW" save code in:
https://github.com/taisel/GameBoy-Online/blob/master/js/other/gui.js#L490

.sav files are usually a binary dump of the SRAM, except RTC (only thing incompatible/wouldn't be cross-emulator).

**Comment by taisel** [comment]
The same save system was also replicated into IodineGBA, although relicensed at MIT, which I need to do for gameboy-online.

**Comment by taisel** [comment]
Since you're contributing with these pull requests, you probably should pull request a copyright update maybe. I changed it to MIT license just now before these pull requests become of considerable size, so the re-licensing of my portion is done before more is pulled in.

**Comment by taisel** [comment]
I really don't care what people do with my contributions, and I don't want to deal with the emails that ask whether it's ok to modify + host at x/y/z due to GPL.

**Comment by taisel** [comment]
https://github.com/taisel/GameBoy-Online/commit/15b26df31846440669114ea43521246f690c4a64

**Comment by jbmorley** [comment]
@taisel Thanks so much for merging this.

As for the license, I plan to make a similar change to gameplay since I would typically choose to use the MIT license.

**Comment by davson19** [comment]
Any updates on this?

**Comment by gradapin** [comment]
It would be really cool !
I tried to figure out where savefiles are stored by looking into the code but couldn't find it.

**Comment by jbmorley** [comment]
Unfortunately .sav files aren't used internally, which is what makes this a separate piece of work. I agree it wold be really cool 👍 .

**Comment by Olenka1810** [comment]
+1 If your read the roms from google drive, maybe sav files (when available) can be dumped there as well? That would be totally awesome :-)

Btw great job on the emulator. Best solution for ios I have seen so far!

**Comment by jbmorley** [comment]
@Olenka1810 That's how I'd like for it to work, yes.

**Comment by welcoMattic** [comment]
I just wrote a small gist to export .sav file from Gameplay local data : https://gist.github.com/welcoMattic/2a2909e5b98aa8d745dfd0df0a6b3091 

Enjoy ;)

**Comment by welcoMattic** [comment]
It can be a base to build the "Export .sav file to Google Drive" feature ;)

**Comment by jbmorley** [comment]
@welcoMattic Thanks for looking into this; that's a great start. I'm looking for ways to test the validity of the save file against different emulators, so if you have any suggestions of emulators you've tested with then that'd be super-helpful.

**Comment by welcoMattic** [comment]
@jbmorley You can test gamebattle, it's used by OpenEmu.
I'm trying to write a small import script, based on https://github.com/taisel/GameBoy-Online source code

**Comment by welcoMattic** [comment]
Ref. #72

**Comment by Shawnjoseph2001** [comment]
It would be helpful if you could at least use Game Play Color saves as backups/ across devices with Game Play Color or Gameboy-Online with raw save data, even if it's not cross platform/emulator.

**Comment by Shawnjoseph2001** [comment]
@welcoMattic Just wondering, how would I use the [this](https://gist.github.com/welcoMattic/2a2909e5b98aa8d745dfd0df0a6b3091) with a copy of Game Play Color/how would I patch it to Game Play Color?

**Comment by welcoMattic** [comment]
@Shawnjoseph2001 I didn't include the Gist into GPC, but juste execute it in Safari web console with my iPhone connected via USB. Hugly hack to extract my save

**Comment by ughwire** [comment]
@welcoMattic I tried executing the code in the safair web console but am getting this error:

![image](https://cloud.githubusercontent.com/assets/15861132/20485065/f7e95550-afc7-11e6-9e88-b84011463f09.png)

Any ideas?

**Comment by duckboy2001** [comment]
Sorry, I'm not too sure what a .sav file is. Will this allow me to save my games to my Google Drive? I just upgraded my phone and Gameplay Color wasn't saved on my iTunes backup. If this worked would I be able to transfer my saved data onto my new phone? I will continue playing on my old phone anyways, but if I didn't have to haul around two phones that would be great.

**Comment by Nythepegasus** [comment]
@duckboy2001 A .sav file is a save file. It would be saved into Google Drive. It would allow to transfer your progress between devices.

**Comment by welcoMattic** [comment]
@ughwire If you don't know Javascript, I recommend you to don't use a Javascript script. 
First, you didn't replace the value `'ROM_NAME'`, so the script can't work.

**Comment by MarceloZeri** [comment]
@welcoMattic Hi, could you help me running your code too? I'm 100% ignorant on js. First, should I escape the spaces in the ROM name? I changed the rom name and the error I'm getting is this one: ReferenceError: Can't find variable: base64ToArray. What do I need to do to have this base64ToArray?
Thanks a lot!

**Comment by MarceloZeri** [comment]
@welcoMattic Hi, could you help me? I managed to run your script. It took me some time to figure out the right name for the rom. Now I get the link at the end but nothing happens when I click it. I know it was supposed to download the contents of escape(rawData). I tried to copy and paste the contents of that variable into a txt file, naming it with the rom name and extension .sav but it is not working. The emulator does not recognize it. I'm using OpenEmu.

I tried also copying the contents of rawData, which results in a file with similar size of a sav file (34k), but had no success either.

<img width="857" alt="screen shot 2017-01-30 at 14 02 19" src="https://cloud.githubusercontent.com/assets/4092290/22431499/d6f5831a-e6f8-11e6-9fb0-123d7d752fc1.png">

**Comment by zensir** [comment]
can anyone write nstroctions on how to import export savs? thanx!

**Comment by kdr264** [comment]
Quick question, has there been any progress on this? Not to sound ungrateful, because I absolutely love gameplay. I would offer to help, but I chose the wrong profession and therefore I don't know any code lol I can paypal people though if thats an incentive :D

**Comment by Vinnybonboot** [comment]
Any progress report on this? I love this app, but I'm about to get a phone upgrade, which means I will lose all of my game saves. I'd be pretty bummed if I have to lose all my progress in the games. It'd be great if I could upload the saves to Google Drive or iCloud.

**Comment by zensir** [comment]
any update on this? is there any way we can help decelope this feature? thnx!

**Comment by joshuafinley** [comment]
I would really love this feature too, this emulator is fantastic and I would like a way of transferring the save to a pc emulator to trade pokemon and battle friends. Keep up the good stuff

**Comment by MikeG1990** [comment]
Fair enough.

**Comment by MikeG1990** [comment]
@Shawnjoseph2001 fair enough I guess I'll have to wait and see if someone comes up with a solution.

**Comment by jbmorley** [comment]
Unforutnately I'm unable to work on this at the moment meaning the fature is on-hold indefinitely. @pvinis Is now maintainer on this project, and people are welcome to submit pull requests.

**Comment by kdr264** [comment]
I commented on this thread over ten months ago asking for any update so either the person who made the emulator doesn't see this thread or doesn't care enough to make his work better.

**Comment by MikeG1990** [comment]
Well the emulator is great thanks for the work you managed to do, it a shame though.

**Comment by MikeG1990** [comment]
I’m new to GitHub, I might sound dumb but what’s a pull request?

**Comment by jbmorley** [comment]
@MikeG1990 I’m glad you’re enjoying the emulator. 

Since the code is open source, anyone is free to take it and modify it, adding whatever functionality they wish. A ‘pull request’ is the process by which that functionality is incorporated into the project.

**Comment by MikeG1990** [comment]
Fair enough, thanks for explaining, because it stored locally inside of Safari i the issue. I honestly don't understand javascript well enough to do something like that. I only knew basics to javascript and haven't done anything with javascript in about 10 years. Maybe there's a clever individual among the users that possibly might be able to help.

**Comment by fankimm** [comment]
> @ughwire If you don't know Javascript, I recommend you to don't use a Javascript script.
> First, you didn't replace the value `'ROM_NAME'`, so the script can't work.

pls help god of programming. I replace 'ROM_NAME' as my rom name.
/*1st try*/  var filename = "myRomName" <- it doesn't work. 
/*2nd try*/ var filename = "myRomName.gbc" <- it doesn't work.

**Comment by MarceloZeri** [comment]
You need to use the ROM's internal name. In my case, for Pokemon Gold, the code was like this:

// Getting the data //
//var filename = "Pokemon - Gold Version (USA, Europe).gbc";
var filename = "POKEMON_GLD";
var rawData = base64ToArray(findValue("B64_SRAM_" + filename)).map(function (a) { return String.fromCharCode(a); }).join("");
...

**Comment by fankimm** [comment]
@MarceloZeri sooooo thanks for reply. but i already know about internal rom name even if i modified file name. I’m trying to import into ‘Game play color’. it isnt easy.

**Comment by fankimm** [comment]
@MarceloZeri sooooo thanks for reply. but i already know about internal rom name even if i modified file name. I’m trying to import into ‘Game play color’. it isnt easy

**Comment by fankimm** [comment]
Finally I successfully imported a .sav file into gameplaycolor !! eureka

**Comment by fankimm** [comment]
How to import .sav files to Game Play Color
https://www.youtube.com/watch?v=t56EQPtCA3w&feature=youtu.be

**Comment by fankimm** [comment]
How to import .sav files to Game Play Color
https://www.youtube.com/watch?v=t56EQPtCA3w&feature=youtu.be
How to export .sav file from Game Play Color https://youtu.be/H5MI1CRyL0E

**Comment by cgc** [comment]
@jbmorley Any interest in implementing this still? I think I could code it up. One simple way to get it done might be making a menu for `didLongPressItem` in `library.js` that has options like `Export save`, `Import save`, `Import save from Drive`, `Export save to Drive`, including a `Delete Rom` option to retain existing functionality. If this sounds good to you, I'm happy to start working on it, should be a relatively small change. Main chunks of work are 1) the menu 2) [file exporting](https://github.com/eligrey/FileSaver.js/blob/master/src/FileSaver.js#L79-L106) 3) [file importing](https://forums.meteor.com/t/webkit-on-ios-ignores-trigger-click-on-file-input/29828) 4) New code to `drive.js` to [update files in drive](https://developers.google.com/drive/api/v3/reference/files/update).

Not quite sure how you'd want to handle `.rtc` files. Would be kind of nice to mirror the format used by [gambatte](https://github.com/sinamas/gambatte/blob/master/libgambatte/src/mem/rtc.h), but I'm not 100% sure how that maps onto [GameBoy-Online's format](https://github.com/gameplaycolor/GameBoy-Online/blob/master/js/GameBoyCore.js#L3934-L3957). At least could simply serialize to GameBoy-Online's format.

Doing a more robust syncing system to Google Drive feels a bit tougher, don't think I'd try to aim for that at first. I think the ideal logic would be: Every time data is synced, record metadata (some hash or file mtime). Data is only synced (from iOS to Drive or Drive to iOS) when one of Drive or iOS have the previously synced file (by comparing to that metadata) and the other has an updated file. If both Drive and iOS have an updated file, either 1) prompt the user for what to do, or 2) make a copy of the Drive file then sync the iOS file to Drive. This sync logic would run after saves & other places we check for data from drive.

**Comment by Klevh** [comment]
> @jbmorley Any interest in implementing this still? I think I could code it up. One simple way to get it done might be making a menu for `didLongPressItem` in `library.js` that has options like `Export save`, `Import save`, `Import save from Drive`, `Export save to Drive`, including a `Delete Rom` option to retain existing functionality. If this sounds good to you, I'm happy to start working on it, should be a relatively small change. Main chunks of work are 1) the menu 2) [file exporting](https://github.com/eligrey/FileSaver.js/blob/master/src/FileSaver.js#L79-L106) 3) [file importing](https://forums.meteor.com/t/webkit-on-ios-ignores-trigger-click-on-file-input/29828) 4) New code to `drive.js` to [update files in drive](https://developers.google.com/drive/api/v3/reference/files/update).
> 
> Not quite sure how you'd want to handle `.rtc` files. Would be kind of nice to mirror the format used by [gambatte](https://github.com/sinamas/gambatte/blob/master/libgambatte/src/mem/rtc.h), but I'm not 100% sure how that maps onto [GameBoy-Online's format](https://github.com/gameplaycolor/GameBoy-Online/blob/master/js/GameBoyCore.js#L3934-L3957). At least could simply serialize to GameBoy-Online's format.
> 
> Doing a more robust syncing system to Google Drive feels a bit tougher, don't think I'd try to aim for that at first. I think the ideal logic would be: Every time data is synced, record metadata (some hash or file mtime). Data is only synced (from iOS to Drive or Drive to iOS) when one of Drive or iOS have the previously synced file (by comparing to that metadata) and the other has an updated file. If both Drive and iOS have an updated file, either 1) prompt the user for what to do, or 2) make a copy of the Drive file then sync the iOS file to Drive. This sync logic would run after saves & other places we check for data from drive.

That sounds great! Any news on this? I can help if needed!

**Comment by adchsm** [comment]
Hey there, any movement on this, would like to see this functionality :)

**Comment by AdriSoup** [comment]
any updates?

**Comment by jbmorley** [comment]
Game Play Color is no longer under active development. Closing as won’t fix.
